test/RSocketTests.cpp

```diff
@@ -53,6 +54,58 @@ std::shared_ptr<RSocketClient> makeClient(
   return makeClientAsync(eventBase, port).get();
 }
 
+namespace {
+struct DisconnectedRequestResponseHandler : public rsocket::RSocketResponder {
+  explicit DisconnectedRequestResponseHandler() {}
+
+  yarpl::Reference<yarpl::single::Single<rsocket::Payload>>
+  handleRequestResponse(rsocket::Payload, rsocket::StreamId) override {
+    assert(false);
+    return nullptr;
+  }
+
+  yarpl::Reference<yarpl::flowable::Flowable<rsocket::Payload>>
+  handleRequestStream(rsocket::Payload, rsocket::StreamId) override {
+    assert(false);
+    return nullptr;
+  }
+
+  yarpl::Reference<yarpl::flowable::Flowable<rsocket::Payload>>
+  handleRequestChannel(
+      rsocket::Payload,
+      yarpl::Reference<yarpl::flowable::Flowable<rsocket::Payload>>,
+      rsocket::StreamId) override {
+    assert(false);
+    return nullptr;
+  }
+
+  void handleFireAndForget(rsocket::Payload, rsocket::StreamId) override {
+    assert(false);
+  }
+
+  void handleMetadataPush(std::unique_ptr<folly::IOBuf>) override {
+    assert(false);
+  }
+
+  ~DisconnectedRequestResponseHandler() {}
+};
+}
+
+std::shared_ptr<RSocketClient> makeDisconnectedClient(
+    folly::EventBase* eventBase) {
+  auto server =
+      makeServer(std::make_shared<DisconnectedRequestResponseHandler>());
+
+  auto client = makeClient(eventBase, *server->listeningPort());
+
+  client->disconnect();
+  folly::Baton<> wait_for_disconnect;
+  eventBase->runInEventBaseThread([&] { wait_for_disconnect.post(); });
```

Comment:
Seems like we should make `RSocketClient::disconnect()` return a future.  This /should/ work, but I don't want to see this done for every single disconnect test case.